crates/sui-core/src/authority/authority_store_types.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use move_core_types::language_storage::StructTag;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use serde_with::Bytes;
+use std::convert::{From, Into};
+use sui_types::base_types::{ObjectDigest, SequenceNumber, TransactionDigest};
+use sui_types::move_package::MovePackage;
+use sui_types::object::{Data, MoveObject, Object, Owner};
+
+/// Forked version of [`sui_types::object::Object`]
+/// Used for efficient storing of move objects in the database
+#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
+pub struct StoreObject {
+    pub data: StoreData,
+    pub owner: Owner,
+    pub previous_transaction: TransactionDigest,
+    pub storage_rebate: u64,
+}
+
+/// Forked version of [`sui_types::object::Data`]
+/// Adds extra enum value `MoveMetadata`, which represents a reference to an object stored separately
+#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
+pub enum StoreData {
+    Move(MoveObject),
+    Package(MovePackage),
+    MoveMetadata(MoveMetadata),
+}
+
+/// Metadata of stored moved object
+#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
+pub struct MoveMetadata {
+    version: SequenceNumber,
+    digest: ObjectDigest,
+}
+
+/// Separately stored move object
+#[serde_as]
+#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
+pub struct StoreMoveObject {
```

Comment:
Could StoreMoveObject just be a tuple of `(MoveObject, u64)`? It seems that other than the refcount the body of StoreMoveObject is the same as MoveObject.

Reply:
MoveObject also contains version number that we don't include in StoreMoveObject

Reply:
actually, one more thing i just realized (which you probably already know) - you won't be able to update the refcount without reading and writing the entire object, which erases any I/O benefit. And, since we can't immediately delete previous object versions (need to support reverts), we would have to increment the refcount even in the case of a transfer or non-mutating read.

Reply:
maybe setting custom merge operator might help in this case (cc: @sadhansood )
FWIU merge is done on rocksdb side, won't require full serialization/deserialization and multiple updates can be stacked
from the docs:
```
RocksDB will more-or-less lazily decide when to actually apply the operations
(e.g.: the next time the user calls Get(), or when the system decides to do its clean-up process called "Compaction").
This means that, when the MergeOperator is actually invoked, it may have several "stacked" operands that need to be applied
```